src/vcpkg/binarycaching.cpp

```diff
@@ -903,179 +906,490 @@ namespace
-                switch (result)
+                provider->push_success(paths, action);
+            }
+
+            m_status[*abi].mark_restored();
+        }
+    }
+
+    static std::vector<CacheStatus*> build_cache_status_vector(
+        const std::vector<Dependencies::InstallPlanAction>& actions,
+        std::unordered_map<std::string, CacheStatus>& status)
+    {
+        std::vector<CacheStatus*> results{actions.size()};
+        for (size_t idx = 0; idx < actions.size(); ++idx)
+        {
+            const auto abi = actions[idx].package_abi().get();
+            if (abi)
+            {
+                results[idx] = &status[*abi];
+            }
+        }
+
+        return results;
+    }
+
+    void BinaryCache::prefetch(const VcpkgPaths& paths, const std::vector<Dependencies::InstallPlanAction>& actions)
+    {
+        auto cache_status = build_cache_status_vector(actions, m_status);
+        for (auto&& provider : m_providers)
+        {
+            provider->prefetch(paths, actions, cache_status);
+            for (auto status : cache_status)
+            {
+                if (status)
                 {
-                    case RestoreResult::build_failed:
-                    case RestoreResult::success: return result;
-                    case RestoreResult::missing: continue;
-                    default: Checks::unreachable(VCPKG_LINE_INFO);
+                    status->mark_unavailable(provider.get());
                 }
             }
-            return RestoreResult::missing;
         }
-        void push_success(const VcpkgPaths& paths, const Dependencies::InstallPlanAction& action) override
+    }
+
+    std::vector<CacheAvailability> BinaryCache::precheck(const VcpkgPaths& paths,
+                                                         const std::vector<Dependencies::InstallPlanAction>& actions)
+    {
+        auto cache_status = build_cache_status_vector(actions, m_status);
+        for (auto&& provider : m_providers)
         {
-            for (auto&& provider : m_providers)
+            provider->precheck(paths, actions, cache_status);
+        }
+
+        std::vector<CacheAvailability> results{actions.size()};
+        for (size_t idx = 0; idx < results.size(); ++idx)
+        {
+            results[idx] = cache_status[idx]->get_available_provider() ? CacheAvailability::available
+                                                                       : CacheAvailability::unavailable;
+        }
+
+        return results;
+    }
+
+    CacheStatus::CacheStatus() noexcept : m_status(CacheStatusState::unknown), m_known_unavailable_providers() { }
+
+    CacheStatus::CacheStatus(const CacheStatus& other) : m_status(other.m_status)
+    {
+        switch (other.m_status)
+        {
+            case CacheStatusState::unknown:
+                ::new (&m_known_unavailable_providers)
+                    std::vector<const IBinaryProvider*>(other.m_known_unavailable_providers);
```

Comment:
Why does this use `::new` (as opposed to `new`)?

Reply:
Standard library habits .... although something about always using that form for 'True Placement New' has always made sense to me....

Although I haven't bothered with the `static_cast<void*>` part outside of standard library code.